[WIP][Distributed] Resolving distributed actor protocols (#2277)
apple/swift-evolution: proposals/NNNN-resolve-distributed-actor-protocols.md

        SerializationRequirement == ActorSystem.SerializationRequirement {
          
  static func resolve(id: ID, using system: ActorSystem) throws -> Self
}
```

This method is not possible to call "on" a protocol since Swift does permit static method calls on protocols in general.

Let us consider a `Greeter` protocol that is constrained to DistributedActor, as follows:

```swift
import Distributed
import SampleDistributed // provides SampleSystem

public protocol Greeter: DistributedActor {
  distributed func greet()
}
```

Swift does not permit static functions to be invoked on a protocol metatype, so sadly we cannot spell this API as follows:

```swift
let system: SampleActorSystem
let greeter = try Greeter.resolve(id: id, using: system)
// ❌ error: static member 'resolve' cannot be used on protocol metatype '(any Greeter).Type'
```

In reality, there is a *synthesized* stub type that is used to implement the specific pairing of this protocol and the specific distributed actor type. However, the purpose of distributed actors is to not leak such runtime details into the language, and we'd like to be able to hide the fact that we're using some form of stub here entirely.

> Therefore proposal introduces makes use of a pattern in Swift which allows the following invocation:

Reviewer: ```suggestion
Therefore, what this proposal introduces makes use of a pattern in Swift which allows the following invocation:
```

Author: Thanks for the fixes Remy!

- introduce the `@DistributedProtocol` macro that can be attached to distributed actor protocols, and enable the use of `resolve(id:using:)` with such types,
- allow distributed actors to be generic over their `ActorSystem`
- extend the distributed metadata section such that a distributed method which is witness to a distributed requirement, is also recorded using the protocol method's `RemoteCallTarget.identifier` and not only the concrete method's

## Proposed solution

### The `@DistributedProtocol` macro

At the core of this proposal is the `@DistributedProtocol` macro. It is an attached declaration macro, which introduces a number of declarations which allow the protocol, or rather, a "stub" type for the protocol, to be used on a client without knowledge about the server implementation's concrete distributed actor type.

The macro must be attached to the a `protocol` declaration that is a `DistributedActor` constrained protocol, like this:

```swift
import Distributed 

@DistributedProtocol
protocol Greeter where ActorSystem: DistributedActorSystem<any Codable> {
  distributed func greet(name: String) -> String
}
```

> The protocol must specify the a constraint on the `ActorSystem` in which it specifies the kind of `SerializationRequirement` it is able to work with. This serialization requirement must be a protocol, and existing distributed actor functionality already will be verifying this.

Reviewer 2: ```suggestion
The protocol must specify a constraint on the `ActorSystem` which specifies the kind of `SerializationRequirement` it is able to work with. This serialization requirement must be a protocol, and existing distributed actor functionality already will be verifying this.
```

```swift
// before: 
// protocol DistributedActorSystem: Sendable {
//   associatedtype SerializationRequirement where ...
//   // ...
// }

// now: 
protocol DistributedActorSystem<SerializationRequirement>: Sendable {
  /// The serialization requirement that will be applied to all distributed targets used with this system.
  associatedtype SerializationRequirement
    where SerializationRequirement == InvocationEncoder.SerializationRequirement,
          SerializationRequirement == InvocationDecoder.SerializationRequirement,
          SerializationRequirement == ResultHandler.SerializationRequirement
  // ...
}

// 
```

The `SerializationRequirement` must be specified for all actors and protocols attempting to abstract over an actor system, because it is necessary to compile-time guarantee the correctness of values passed to such distributed actor methods. The compiler uses this associated type to verify all argument types and returned values are able to be serialized when performing remote calls, and will refuse to compile invocations would otherwise would have failed at runtime.

> Thanks to this new primary associated type, it is now possible to spell such our "distributed async sequence" as a generic actor, implement it once, and re-use it across any compatible actor system implementation:

Reviewer 2: ```suggestion
Thanks to this new primary associated type, it is now possible to spell our `DistributedAsyncSequence` as a generic actor, implement it once, and re-use it across any compatible actor system implementation:
```

```swift
distributed actor DistributedAsyncSequence<Element, ActorSystem> 
  where Element: Sendable & Codable,
        ActorSystem: DistributedActorSystem<any Codable> { 
          
  // not real implementation; simplified method to showcase introduced capabilities
  distributed func exampleNextElement() async throws -> Element? { ... }
}
```

Note that since the `ActorSystem` specifies a concrete `SerializationRequirement` the compiler is still able to check that all types invoked in a distributed function call conform to this protocol, i.e. we're guaranteed to be able to serialize `Element` because the ActorSystem provided must be able to handle this serialization mechanism.

This also extends to distributed protocols, which are now able to abstract over an actor system, while specifying what serialization requirement they support:

```swift
protocol DistributedAsyncSequence: DistributedActor 
    where ActorSystem: DistributedActorSystem<any Codable> {
  associatedtype Element: Sendable & Codable
      
  distributed func exampleNextElement() async throws -> Element? { ... }
}
```

> Missing to specify the serialization requirement is a compile time error:

Reviewer 2: ```suggestion
Failing to specify the serialization requirement is a compile time error:
```

```swift
protocol DistributedAsyncSequence: DistributedActor 
    where ActorSystem: DistributedActorSystem { 
    // error: distributed actor protocol must specify `ActorSystem.SerializationRequirement`,
    // you can provide it like this: DistributedActorSystem<any Codable>
  associatedtype Element: Sendable & Codable
      
  distributed func exampleNextElement() async throws -> Element? { ... }
}
```

The serialization requirement must be a `protocol` type; This was previously enforced, and remains so after this proposal. The important part is that the macro is able to synthesize an stub implementation type, that the existing resolution mechanisms can be invoked on. 

## Detailed design

The `@DistributedProtocol` macro generates a concrete `distributed actor` declaration as well as an extension which implements the protocol's method requirements with "stub" implementations.

> >  **NOTE:** The exact details of the macro synthesized code are not guaranteed to remain the same, and may change without notice.

Reviewer 2: With this note here is it perhaps worth also being very explicit about what _is_ guaranteed to remain the same? I realize this is distributed throughout the proposal an implicit in Swift's source compatibility guarantee, but if we're adding this scary note I think it would be good to indicate what users _can_ actually rely on.

Author: Sounds good -- I think we will guarantee the existence of the `$Greeter` for a `protocol Greeter`, I'll call it out

> >  **NOTE:** The exact details of the macro synthesized code are not guaranteed to remain the same, and may change without notice.

Author: ```suggestion
>  **NOTE:** The exact details of the macro synthesized code are not guaranteed to remain the same, and may change without notice. The existence of the $-prefixed generated type is guaranteed however, as it is the public API how developers resolve and obtain remote references, I.e. for a `protocol Greeter` annotated with the `@DistributedProtocol` macro, developers may rely on the existence of the `distributed actor $Greeter` with the same access level as the protocol.
```

This proposal also introduces an empty `DistributedActorStub` protocol:

```swift
public protocol DistributedActorStub where Self: DistributedActor {}
```

> The `@DistributedProtocol` macro synthesizes a concrete distributed actor which accepts a generic `ActorSystem`. The generated actor declaration matches access level with the original declaration, and implements the protocol as well as the `DistributedStub` protocol:

Reviewer 2: ```suggestion
The `@DistributedProtocol` macro synthesizes a concrete distributed actor which accepts a generic `ActorSystem`. The generated actor declaration matches access level with the original declaration, and implements the protocol as well as the `DistributedActorStub` protocol:
```

}

// "stub" type
distributed actor $Greeter<ActorSystem>: Greeter, DistributedStubActor
    where DistributedActorSystem<any Codable> {
  private init() {} // cannot initialize, can only resolve(id:using:)
}

extension Greeter where Self: DistributedActorStub {
  // ... stub implementations for protocol requirements ...
}
```

Default implementations for all the protocol's requirements (including non-distributed requirements) are provided by extensions utilizing the `DistributedActorStub` protocol.

It is possible for a protocol type to inherit other protocols, in that case the parent protocol must either have default implementations for all its requirements, or it must also apply the `@DistributedProtocol` protocol which generates such default implementations.

The default method "stub" implementations provided by the `@DistributedProtocol` simply fatal error if they were to ever be invoked. In practice, invoking those methods is not possible, because resolving a stub will always return a remote reference, and therefore calls on these methods are redirected to `DistributedActorSystem`'s `remoteCall` rather than invoking the "local" methods.  

> It is recommended to use `some/any Greeter` rather than `$Greeter` when passing around resolved instances of a distributed actor protocol. This way none of your code is tied to the fact of using a specific type of proxy, but rather, can accept any greeter, be it local or resolved through a proxy reference. This can come in handy when refactoring a codebase, and merging modules in such way where the greeter may actually be a local instance in some situations.

Reviewer 2: ```suggestion
It is recommended to use `some Greeter` or `any Greeter` rather than `$Greeter` when passing around resolved instances of a distributed actor protocol. This way none of your code is tied to the fact of using a specific type of proxy, but rather, can accept any greeter, be it local or resolved through a proxy reference. This can come in handy when refactoring a codebase, and merging modules in such way where the greeter may actually be a local instance in some situations.
```

### Interaction with the `DefaultDistributedActorSystem`

> Since the introduction of distributed actors, it is possible to declare a module wide `DefaultDistributedActorSystem` type alias, like this:

Reviewer 2: ```suggestion
Since the introduction of distributed actors, it is possible to declare a module-wide `DefaultDistributedActorSystem` type alias, like this:
```

```swift
typealias DefaultDistributedActorSystem = ClusterSystem
```

This makes it easier to declare distributed actors as the `ActorSystem` type requirement is witnessed by an implicit type alias generated in every concrete distributed actor, like this:

```swift
distributed actor Worker { 
  // synthesized:
  // typealias ActorSystem = ClusterSystem // because 'DefaultDistributedActorSystem = ClusterSystem'
  
  distributed func work()
}
```

The newly introduced ability to abstract over the `ActorSystem` in concrete distributed actors _wins_ over the synthesized typealias, causing the typealias to not be emitted:

```swift
distributed actor Worker<ActorSystem> where ActorSystem: DistributedActorSystem<any Codable> {
  distributed func work()
}
```

The `ActorSystem` type requirement of the `DistributedActorProtocol` in this case is witnessed by the generic parameter, and not by the "default" fallback type.

> This is the right behavior because this generic worked can work on _any_ distributed actor system where the `SerializationRequirement` is Codable, and not only on the `ClusterSystem`.

Reviewer 2: ```suggestion
This is the right behavior because this generic type works with _any_ distributed actor system where the `SerializationRequirement` is Codable, and not only on the `ClusterSystem`.
```

### Extend Distributed metadata for protocol method identifier lookups

The way distributed method invocations work on the recipient node is that a message is parsed from some incoming transport, and a `RemoteCallTarget` is recovered. The remote call target in currently is a mangled encoding of the concrete distributed method the call was made for, like this:

A shared module introducing the `Capybara` protocol:

```swift
// Module "Shared"
@DistributedProtocol
public protocol Capybara where ActorSystem: DistributedActorSystem<any Codable> {
  distributed var name: String { get }
  distributed func eat()
}
```

The distributed actor runtime stores static metadata about distributed methods, such that the `executeDistributedTarget(on:target:invocationDecoder:handler:)` method is able to turn the mangled `RemoteCallIdentifier` into a concrete method handle that it then invokes. This proposal introduces a way to mangle calls on distributed protocol requirements, in such a way that they refer to the `$`-prefixed name, and invocations on such accessor are performed on the recipient side using the concrete actor's witness tables.

> We can illustrate the new remoteCall flow like this:

Reviewer 2: ```suggestion
We can illustrate the new `remoteCall` flow like this:
```

The client-side does not need to know or care about what concrete implementation type is used to implement this call on the recipient system, as it will only ever be performing such distributed protocol based calls.

**Recipient, i.e. server side:**

```swift
// Module MyActorSystem

final class MySampleDistributedActorSystem: DistributedActorSystem, ... {
  // ... 
  
  func findById(_ id: ActorID) -> (any DistributedActor)? { ... }
  
  func receiveMessage() async throws {
    let envelope: MyTransportEnvelope = try await readFromNetwork()
    
    guard let actor = findById(envelope.id) else {
      throw TargetActorNotFound(envelope.id)
    }

    // RemoteCallIdentifier("Shared.Capybara.name") <<< 
    let target: RemoteCallTarget = envelope.target 
    try await executeDistributedTarget(
      on: actor,
      target: target, // the protocol method identifier
      invocationDecoder: invocation.makeDecoder(),
      handler: resultHandler
    }
)
```

> This logic is exactly the same as any existing `DistributedActorSystem` implementation -- however changes in the Distributed runtime will handle the protocol method invocation and be able to route it to the concrete resolved actor (returned by `findByID` in this snippet, e.g. the `Caplin` concrete type).

Reviewer 2: ```suggestion
This logic is exactly the same as any existing `DistributedActorSystem` implementation -- however, changes in the Distributed runtime will handle the protocol method invocation and be able to route it to the concrete resolved actor (returned by `findByID` in this snippet, e.g. the `Caplin` concrete type).
```

let g: any Greeter = try .resolve(id: id, using: someSystem)
```

This would have to synthesize an ad-hoc created anonymous declaration for a `$Greeter` and at the site of the `resolve` type-check if the declaration can be used with the `someSystem`'s serialization requirement. We would have to check if the distributed greet method's parameters and return type conform to `Codable` etc, and all this would have to happen lazily -- triggered by the existence of a `.resolve` method combining a protocol with a specific actor system.

The only possible spelling of such API would have been this: `let g: any Greeter = try .resolve(...)` as the concrete type that is used to implement this `any Greeter` is not user visible, and cannot be. This is a lot of complexity, to what amounts to just a simple stub type.

We believe that the macro stub approach is a good balance between convenience and lack of "magic" compiler support, as for this specific piece of the design no deep integration in the type system is necessary.

## Revisions

- 1.2
  - Change implementation to not need `#resolve` macro, but rely on generic distributed actors
  - General cleanup

> - 2.1

Reviewer 3: Seemed typo here?

> - 2.1

Author: ```suggestion
```
